### ISSUE TITLE
[BUG] Unable to merge `IndexedChoiceMap` and `VectorChoiceMap` when shapes are compatible

### ISSUE THREAD

**Issue by ahiser1117** [genjax-community/genjax#959]
**Describe the bug**
I have an situation where I want to merge an `IndexedChoiceMap` with a `VectorChoiceMap`. They have the same general shape and I am able to get the merge to work as expected if I first convert the `IndexedChoiceMap` into a `VectorChoiceMap`. This is not possible in the situation I am trying to make this work so I need the merge to work.

**To Reproduce**
I have this `HierarchicalChoiceMap` defined which contains an `IndexedChoiceMap`:
```python
console.print(provided)
```
```
(HierarchicalChoiceMap)
├── :σ0_noise
│   └── (ValueChoice)
│       └──  f32[]
├── :c_vT
│   └── (ValueChoice)
│       └──  f32[]
├── :c_P
│   └── (ValueChoice)
│       └──  f32[]
├── :c_v
│   └── (ValueChoice)
│       └──  f32[]
├── :α0
│   └── (ValueChoice)
│       └──  f32[]
├── :s0
│   └── (ValueChoice)
│       └──  f32[]
├── :σ0_SE
│   └── (ValueChoice)
│       └──  f32[]
├── :ys
│   └── (IndexedChoiceMap,  i32[38])
│       └── (HierarchicalChoiceMap)
│           ├── :b
│           │   └── (ValueChoice)
│           │       └──  f32[38]
│           ├── :ζ0
│           │   └── (ValueChoice)
│           │       └──  f32[38]
│           └── :y0
│               └── (ValueChoice)
│                   └──  f32[38]
└── :c_θh
    └── (ValueChoice)
        └──  f32[]
```
I want to merge it with another `HierarchicalChoiceMap` that contains a `VectorChoiceMap`:
```python
console.print(fixed)
```
```
(HierarchicalChoiceMap)
└── :ys
    └── (VectorChoiceMap)
        └── (HierarchicalChoiceMap)
            └── :y
                └── (ValueChoice)
                    └──  f32[38,1600]
```
Trying to merge these results in an error:
```python
merged = fixed.safe_merge(provided)
```
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[87], line 1
----> 1 merged = fixed.safe_merge(provided)

    [... skipping hidden 1 frame]

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/genjax/_src/core/datatypes/generative.py:243, in Choice.safe_merge(self, other)
    242 def safe_merge(self, other: "Choice") -> "Choice":
--> 243     new, discard = self.merge(other)
    245     # If the discarded choice is not empty, raise an error.
    246     # However, it's possible that we don't know that the discarded
    247     # choice is empty until runtime, so we use checkify.
    248     def _check():

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/plum/function.py:484, in _BoundFunction.__call__(self, _, *args, **kw_args)
    483 def __call__(self, _, *args, **kw_args):
--> 484     return self._f(self._instance, *args, **kw_args)

    [... skipping hidden 1 frame]

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/genjax/_src/core/datatypes/generative.py:1352, in HierarchicalChoiceMap.merge(self, other)
   1350 if other.has_submap(k):
   1351     sub = other.get_submap(k)
-> 1352     new[k], discard[k] = v.merge(sub)
   1353 else:
   1354     new[k] = v

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/plum/function.py:484, in _BoundFunction.__call__(self, _, *args, **kw_args)
    483 def __call__(self, _, *args, **kw_args):
--> 484     return self._f(self._instance, *args, **kw_args)

    [... skipping hidden 1 frame]

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:304, in VectorChoiceMap.merge(self, other)
    301     return v1.at[indices].set(v2)
    303 assert jtu.tree_structure(self.inner, lambda x: isinstance(x, HashableDict)) == jtu.tree_structure(new, lambda x: isinstance(x, HashableDict))
--> 304 new = jtu.tree_map(_inner, self.inner, new)
    306 return VectorChoiceMap(new), IndexedChoiceMap(indices, discard)

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/jax/_src/tree_util.py:243, in tree_map(f, tree, is_leaf, *rest)
    210 """Maps a multi-input function over pytree args to produce a new pytree.
    211 
    212 Args:
   (...)
    240   [[5, 7, 9], [6, 1, 2]]
    241 """
    242 leaves, treedef = tree_flatten(tree, is_leaf)
--> 243 all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]
    244 return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/jax/_src/tree_util.py:243, in <listcomp>(.0)
    210 """Maps a multi-input function over pytree args to produce a new pytree.
    211 
    212 Args:
   (...)
    240   [[5, 7, 9], [6, 1, 2]]
    241 """
    242 leaves, treedef = tree_flatten(tree, is_leaf)
--> 243 all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]
    244 return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))

ValueError: Mismatch custom node data: ['y'] != ['b', 'y', 'ζ0', 'y0']; value: {'y': ChoiceValue(value=f32[38,1600]), 'b': ChoiceValue(value=f32[38]), 'ζ0': ChoiceValue(value=f32[38]), 'y0': ChoiceValue(value=f32[38])}.
```

**Expected behavior**
I can hack this to get it to work by translating the `IndexedChoiceMap` back into a `VectorChoiceMap` and get the desired result. Unfortunately, I can't do this in my use case since it is internal.

```python
vc_chm = vector_choice_map(provided['ys'].inner.trie.inner)
new_provided = provided.insert('ys', vc_chm)
console.print(new_provided)
```
```
(HierarchicalChoiceMap)
├── :σ0_noise
│   └── (ValueChoice)
│       └──  f32[]
├── :c_vT
│   └── (ValueChoice)
│       └──  f32[]
├── :c_P
│   └── (ValueChoice)
│       └──  f32[]
├── :c_v
│   └── (ValueChoice)
│       └──  f32[]
├── :α0
│   └── (ValueChoice)
│       └──  f32[]
├── :s0
│   └── (ValueChoice)
│       └──  f32[]
├── :σ0_SE
│   └── (ValueChoice)
│       └──  f32[]
├── :ys
│   └── (VectorChoiceMap)
│       └── (HierarchicalChoiceMap)
│           ├── :b
│           │   └── (ValueChoice)
│           │       └──  f32[38]
│           ├── :ζ0
│           │   └── (ValueChoice)
│           │       └──  f32[38]
│           └── :y0
│               └── (ValueChoice)
│                   └──  f32[38]
└── :c_θh
    └── (ValueChoice)
        └──  f32[]
```
Which then allows me to merge with the `VectorChoiceMap` I wanted in the first place.
```python
merged = fixed.safe_merge(new_provided)
console.print(merged)
```
```
(HierarchicalChoiceMap)
├── :ys
│   └── (VectorChoiceMap)
│       └── (HierarchicalChoiceMap)
│           ├── :y
│           │   └── (ValueChoice)
│           │       └──  f32[38,1600]
│           ├── :b
│           │   └── (ValueChoice)
│           │       └──  f32[38]
│           ├── :ζ0
│           │   └── (ValueChoice)
│           │       └──  f32[38]
│           └── :y0
│               └── (ValueChoice)
│                   └──  f32[38]
├── :σ0_noise
│   └── (ValueChoice)
│       └──  f32[]
├── :c_vT
│   └── (ValueChoice)
│       └──  f32[]
├── :c_P
│   └── (ValueChoice)
│       └──  f32[]
├── :c_v
│   └── (ValueChoice)
│       └──  f32[]
├── :α0
│   └── (ValueChoice)
│       └──  f32[]
├── :s0
│   └── (ValueChoice)
│       └──  f32[]
├── :σ0_SE
│   └── (ValueChoice)
│       └──  f32[]
└── :c_θh
    └── (ValueChoice)
        └──  f32[]
```

**Execution environment (please complete the following information):**
 - OS: Ubuntu 22.04
 - Python version: 3.11.5
 - GenJAX version: 0.2.0

**Comment by femtomc** [comment]
@ahiser1117 Thanks for reporting this!

> Unfortunately, I can't do this in my use case since it is internal.

Does this mean that, in your code -- you can't access the place where the merge is occurring? (Curious, just to know how blocking)

**Comment by ahiser1117** [comment]
@femtomc What I meant by internal is the problem is occurring when I call `NUTS` from `BlackJax`. It's calling `merge` on the `PyTree`s I'm passing to it. Given the current definition of the datastructures, namely `VectorChoiceMap` and `IndexedChoiceMap` it wasn't happy.

I have been able to get it to work by defining a `VectorSelection` datatype that is created when `get_selection()` is called on the `VectorChoiceMap`, so an `IndexedChoiceMap` is never created. The `merge` function then worked fine and I was able to call the `NUTS` implementation from `BlackJax`
